src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -5155,273 +5157,413 @@ MergeTreeData::PartsBackupEntries MergeTreeData::backupParts(
-    auto restored_parts_holder = std::make_shared<RestoredPartsHolder>(std::static_pointer_cast<MergeTreeData>(shared_from_this()), backup);
+            if (!retryable || (try_no + 1 == loading_parts_max_tries))
+            {
+                if (Exception * e = exception_cast<Exception *>(error))
+                    e->addMessage("while restoring part {} of table {}", part->name, storage.getStorageID());
+                std::rethrow_exception(error);
+            }
 
-    fs::path data_path_in_backup_fs = data_path_in_backup;
-    size_t num_parts = 0;
+            tryLogException(error, &Poco::Logger::get(storage.getLogName()),
+                            fmt::format("Failed to load part {} at try {} with a retryable error. Will retry in {} ms", part_name, try_no, backoff_ms));
 
-    for (const String & part_name : part_names)
-    {
-        const auto part_info = MergeTreePartInfo::tryParsePartName(part_name, format_version);
-        if (!part_info)
-        {
-            throw Exception(ErrorCodes::CANNOT_RESTORE_TABLE, "File name {} is not a part's name",
-                            String{data_path_in_backup_fs / part_name});
+            std::this_thread::sleep_for(std::chrono::milliseconds(backoff_ms));
+            backoff_ms = std::min(backoff_ms * 2, loading_parts_max_backoff_ms);
         }
 
-        if (partition_ids && !partition_ids->contains(part_info->partition_id))
-            continue;
+        UNREACHABLE();
+    }
 
-        restorer.addDataRestoreTask(
-            [storage = std::static_pointer_cast<MergeTreeData>(shared_from_this()),
-             backup,
-             part_path_in_backup = data_path_in_backup_fs / part_name,
-             my_part_info = *part_info,
-             restore_broken_parts_as_detached,
-             restored_parts_holder]
-            { storage->restorePartFromBackup(restored_parts_holder, my_part_info, part_path_in_backup, restore_broken_parts_as_detached); });
+    /// Parts from a backup are collected before attaching them to the table to preserve the same order of parts.
+    class PartsFromBackupCollector
+    {
+    public:
+        void addPart(std::shared_ptr<IMergeTreeDataPart> && part)
+        {
+            std::lock_guard lock{mutex};
+            parts.emplace_back(std::move(part));
+        }
 
-        ++num_parts;
-    }
+        /// Returns a list of all added parts sorted by `min_block`.
+        /// (The list is sorted because we need to preserve the order of parts.)
+        std::vector<std::shared_ptr<IMergeTreeDataPart>> getParts()
+        {
+            std::lock_guard lock{mutex};
+            std::vector<std::shared_ptr<IMergeTreeDataPart>> res;
+            std::swap(parts, res);
+            auto less_by_min_block = [](const std::shared_ptr<IMergeTreeDataPart> & left, const std::shared_ptr<IMergeTreeDataPart> & right)
+            { return left->info.min_block < right->info.min_block; };
+            std::sort(res.begin(), res.end(), less_by_min_block);
```

Comment:
It's the `MergeTreeData` class - the common code for `MergeTree` and `ReplicatedMergeTree`. For `MergeTree` it's important to sort parts by `min_block`, for `ReplicatedMergeTree` we can either sort them by `min_block` or use `MergeTreePartInfo::operator <` (which is the same `LessDataPart`). There is no harm if we sort parts by `min_block` for `ReplicatedMergeTree` and then attach them - the order of parts within any partition will be correct.

---

Anyway there is the setting [restore_allocates_block_numbers_in_batch](https://github.com/ClickHouse/ClickHouse/blob/8bb2f0680ab767b82762a7403a96e9b0f70dd206/src/Core/ServerSettings.h#L41) which I'm adding in this PR:

if it's true then the functions [calculateBlockNumbersForRestoring*()](https://github.com/ClickHouse/ClickHouse/blob/8bb2f0680ab767b82762a7403a96e9b0f70dd206/src/Storages/MergeTree/calculateBlockNumbersForRestoring.h) are used,
and this class [PartsFromBackupCollector](https://github.com/ClickHouse/ClickHouse/blob/8bb2f0680ab767b82762a7403a96e9b0f70dd206/src/Storages/MergeTree/MergeTreeData.cpp#L5365) is not used. 

if it's false then the functions [calculateBlockNumbersForRestoring*()](https://github.com/ClickHouse/ClickHouse/blob/8bb2f0680ab767b82762a7403a96e9b0f70dd206/src/Storages/MergeTree/calculateBlockNumbersForRestoring.h) are not used,
and this class [PartsFromBackupCollector](https://github.com/ClickHouse/ClickHouse/blob/8bb2f0680ab767b82762a7403a96e9b0f70dd206/src/Storages/MergeTree/MergeTreeData.cpp#L5365) is used.

By default `restore_allocates_block_numbers_in_batch` is true. So this class represents the old behaviour which I think it's better to keep in the code for a while in case something goes wrong.